deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,257 @@
+        """
+        ...
+
+    def connect(self, sender: str, receiver: str) -> "PipelineProtocol":
+        """Connect two components in the pipeline.
+
+        :param sender: Component name or component.connection_name that sends data
+        :param receiver: Component name or component.connection_name that receives data
+        :return: The pipeline instance for method chaining
+        """
+        ...
+
+
+class PipelineService:
+    """Handles the publishing of Haystack pipelines and indexes to deepset Cloud."""
+
+    def __init__(self, api: DeepsetCloudAPI) -> None:
+        """Initialize the pipeline service.
+
+        :param api: An initialized DeepsetCloudAPI instance
+        """
+        self.api = api
+
+    @classmethod
+    async def factory(cls, config: CommonConfig) -> "PipelineService":
+        """Create a new instance of the pipeline service.
+
+        :param config: CommonConfig object.
+        """
+        async with DeepsetCloudAPI.factory(config) as api:
+            return cls(api)
+
+    def _append_section_to_yaml(self, yaml_str: str, section_name: str, fields: dict) -> str:
+        """Append a section configuration to the YAML string.
+
+        :param yaml_str: Original YAML string
+        :param section_name: Name of the section to append (e.g., 'inputs' or 'outputs')
+        :param fields: Dictionary of fields to append
+        :return: YAML string with section appended
+        """
+        # Remove empty values
+        fields = {k: v for k, v in fields.items() if v}
+
+        if not fields:
+            return yaml_str
+
+        section_yaml = f"\n{section_name}:\n"
+        for key, value in fields.items():
+            if isinstance(value, list):
+                section_yaml += f"  {key}:\n"
+                for item in value:
+                    section_yaml += f"  - {item}\n"
+            else:
+                section_yaml += f"  {key}: {value}\n"
+
+        return yaml_str.rstrip() + "\n" + section_yaml
+
+    def _append_inputs_to_yaml(self, yaml_str: str, inputs: PipelineInputs) -> str:
+        """Append inputs configuration to the YAML string.
+
+        :param yaml_str: Original YAML string
+        :param inputs: Input configuration to append
+        :return: YAML string with inputs appended
+        """
+        input_fields = inputs.model_dump()
+        return self._append_section_to_yaml(yaml_str, "inputs", input_fields)
+
+    def _append_outputs_to_yaml(self, yaml_str: str, outputs: PipelineOutputs) -> str:
+        """Append outputs configuration to the YAML string.
+
+        :param yaml_str: Original YAML string
+        :param outputs: Output configuration to append
+        :return: YAML string with outputs appended
+        """
+        output_fields = outputs.model_dump()
+        return self._append_section_to_yaml(yaml_str, "outputs", output_fields)
+
+    def _create_config_yaml(self, pipeline: PipelineProtocol, config: PublishConfig) -> str:
```

Comment:
nit/ general style thing: I usually find it easier to read code top to bottom. So scrolling down to get more detailed instead of scrolling up like it's here.

https://github.com/Geeksltd/Programming.Tips/blob/master/docs/methods/stepdown-rule.md

https://dev.to/aralroca/how-do-you-order-your-functions-45cf